# Khai báo thư viện


In [36]:
import pandas as pd

# Load data

In [37]:
df = pd.read_csv("/content/bike_buyers.csv")

# EDA

In [38]:
df.head()

,ID,Marital Status,Gender,Income,Children,Education,Occupation,Home Owner,Cars,Commute Distance,Region,Age,Purchased Bike
0,12496,M,F,"$40.000,00",1,Bachelors,Skilled Manual,Yes,0,0-1 Miles,Europe,42,No
1,24107,M,M,"$30.000,00",3,Partial College,Clerical,Yes,1,0-1 Miles,Europe,43,No
2,14177,M,M,"$80.000,00",5,Partial College,Professional,No,2,2-5 Miles,Europe,60,No
3,24381,S,M,"$70.000,00",0,Bachelors,Professional,Yes,1,5-10 Miles,Pacific,41,Yes
4,25597,S,M,"$30.000,00",0,Bachelors,Clerical,No,0,0-1 Miles,Europe,36,Yes


In [39]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1026 entries, 0 to 1025
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                1026 non-null   int64 
 1   Marital Status    1026 non-null   object
 2   Gender            1026 non-null   object
 3   Income            1026 non-null   object
 4   Children          1026 non-null   int64 
 5   Education         1026 non-null   object
 6   Occupation        1026 non-null   object
 7   Home Owner        1026 non-null   object
 8   Cars              1026 non-null   int64 
 9   Commute Distance  1026 non-null   object
 10  Region            1026 non-null   object
 11  Age               1026 non-null   int64 
 12  Purchased Bike    1026 non-null   object
dtypes: int64(4), object(9)
memory usage: 104.3+ KB


In [40]:
df.ID.nunique()

1000

In [41]:
df.ID.value_counts()

,count
ID,
16466,2
18299,2
18283,2
17841,2
12590,2
...,...
17238,1
23608,1
22538,1


Trường thông tin "Income" là thu nhập lẽ ra phải là kiểu dữ liệu Int thay vì object => cần xem xét lại.

Đối với cột ID, phần lớn các giá trị là duy nhất, tuy nhiên vẫn tồn tại một số trường hợp trùng lặp. Cần tiến hành rà soát và nhóm các khách hàng có cùng ID lại với nhau, vì mục tiêu của bài toán là phân tích hành vi khách hàng mua xe đạp. Điều này sẽ giúp tránh sai lệch trong kết quả và phản ánh đúng dữ liệu thực tế.

In [42]:
df["Marital Status"].value_counts()

,count
Marital Status,
M,549
S,477


In [43]:
df.Gender.value_counts()

,count
Gender,
M,525
F,501


In [44]:
df.Income.value_counts()

,count
Income,
"$60.000,00",166
"$40.000,00",159
"$30.000,00",140
"$70.000,00",125
"$80.000,00",94
"$20.000,00",78
"$10.000,00",76
"$90.000,00",40
"$50.000,00",40


In [45]:
df.Children.value_counts()

,count
Children,
0,289
2,219
1,172
3,136
4,126
5,84


In [46]:
df.Education.value_counts()

,count
Education,
Bachelors,311
Partial College,278
High School,184
Graduate Degree,175
Partial High School,78


In [47]:
df.Occupation.value_counts()

,count
Occupation,
Professional,280
Skilled Manual,259
Clerical,187
Management,174
Manual,126


In [48]:
df["Home Owner"].value_counts()

,count
Home Owner,
Yes,697
No,329


In [49]:
df.Cars.value_counts()

,count
Cars,
2,348
1,277
0,254
3,86
4,61


In [50]:
df["Commute Distance"].value_counts()

,count
Commute Distance,
0-1 Miles,378
5-10 Miles,197
1-2 Miles,176
2-5 Miles,162
10+ Miles,113


In [51]:
df.Region.value_counts()

,count
Region,
North America,508
Europe,316
Pacific,202


In [52]:
df.Age.value_counts()

,count
Age,
40,44
35,40
47,40
36,39
38,38
43,36
42,34
32,34
45,32


In [53]:
df['Age'].describe()

,Age
count,1026.000000
mean,44.138402
std,11.349282
min,25.000000
25%,35.000000
50%,43.000000
75%,52.000000
max,89.000000


Phạm vi tuổi của khách hàng có vẻ là hợp lý.

# Preprocessing - Tiền xử lý dữ liệu

## Xoá dữ liệu trùng lặp

In [54]:
df = df.drop_duplicates()

In [55]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1000 entries, 0 to 999
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                1000 non-null   int64 
 1   Marital Status    1000 non-null   object
 2   Gender            1000 non-null   object
 3   Income            1000 non-null   object
 4   Children          1000 non-null   int64 
 5   Education         1000 non-null   object
 6   Occupation        1000 non-null   object
 7   Home Owner        1000 non-null   object
 8   Cars              1000 non-null   int64 
 9   Commute Distance  1000 non-null   object
 10  Region            1000 non-null   object
 11  Age               1000 non-null   int64 
 12  Purchased Bike    1000 non-null   object
dtypes: int64(4), object(9)
memory usage: 109.4+ KB


In [56]:
df.ID.value_counts()

,count
ID,
12121,1
12496,1
24107,1
14177,1
24381,1
...,...
11434,1
12697,1
22173,1


Dữ liệu đã giảm xuống từ 1026 thành 1000 dòng (số 1000 đúng bằng thống kê số lượng dữ liệu khác nhau của cột ID). Như vậy vấn đề ID trùng nhau là do dữ liệu bị lặp lại.

## Chuyển cột Income về định dạng dữ liệu Int

In [57]:
df['Income'] = (
    df['Income']
    .str.replace('$', '', regex=False)     # bỏ ký hiệu $
    .str.replace('.', '', regex=False)     # bỏ dấu chấm (phân cách nghìn)
    .str.replace(',', '.', regex=False)    # đổi dấu phẩy thành dấu chấm (thập phân)
    .astype(float)                         # chuyển sang float
    .astype(int)                           # chuyển sang int
)

In [58]:
df.Income.value_counts()

,count
Income,
60000,165
40000,154
30000,135
70000,124
80000,91
20000,74
10000,73
50000,40
90000,39


## Sửa lại giá trị cho cột Marital Status

In [59]:
df['Marital Status'] = df['Marital Status'].map({
    'M': 'Married',
    'S': 'Single'
})

In [60]:
df.head()

,ID,Marital Status,Gender,Income,Children,Education,Occupation,Home Owner,Cars,Commute Distance,Region,Age,Purchased Bike
0,12496,Married,F,40000,1,Bachelors,Skilled Manual,Yes,0,0-1 Miles,Europe,42,No
1,24107,Married,M,30000,3,Partial College,Clerical,Yes,1,0-1 Miles,Europe,43,No
2,14177,Married,M,80000,5,Partial College,Professional,No,2,2-5 Miles,Europe,60,No
3,24381,Single,M,70000,0,Bachelors,Professional,Yes,1,5-10 Miles,Pacific,41,Yes
4,25597,Single,M,30000,0,Bachelors,Clerical,No,0,0-1 Miles,Europe,36,Yes


## Sửa lại giá trị cho cột Gender

In [61]:
df['Gender'] = df['Gender'].replace({
    'M': 'Male',
    'F': 'Female'
})

In [62]:
df.head()

,ID,Marital Status,Gender,Income,Children,Education,Occupation,Home Owner,Cars,Commute Distance,Region,Age,Purchased Bike
0,12496,Married,Female,40000,1,Bachelors,Skilled Manual,Yes,0,0-1 Miles,Europe,42,No
1,24107,Married,Male,30000,3,Partial College,Clerical,Yes,1,0-1 Miles,Europe,43,No
2,14177,Married,Male,80000,5,Partial College,Professional,No,2,2-5 Miles,Europe,60,No
3,24381,Single,Male,70000,0,Bachelors,Professional,Yes,1,5-10 Miles,Pacific,41,Yes
4,25597,Single,Male,30000,0,Bachelors,Clerical,No,0,0-1 Miles,Europe,36,Yes


## Tạo cột phân cụm tuổi tác

In [63]:
df['Age Group'] = df['Age'].apply(
    lambda x: 'Adult(0-30)' if x <= 30
    else 'Middle Age(31-54)' if x <= 54
    else 'Old(54+)'
)

## Thay đổi giá trị trong cột Commute Distance

In [64]:
df['Commute Distance'] = df['Commute Distance'].replace(
    '10+ Miles', 'More than 10 miles '
)

In [65]:
df["Commute Distance"].value_counts()

,count
Commute Distance,
0-1 Miles,366
5-10 Miles,192
1-2 Miles,169
2-5 Miles,162
More than 10 miles,111
